### PR TITLE
Use files on local machine in a pipeline (Experiment)

### PR DESCRIPTION
**Pull request by sm43**
TEP (WIP)

This is an experiment to try out if it is possible to use local source and it seems it is :)

This adds support to use copy files on the local machine to a workspace and use it in a pipeline instead of 
cloning from a git repository.

This is currently implemented in the `tkn pipeline start` command in which if the 
the pipeline has a workspace then it will ask the user if they want to copy files from 
local machine into the workspace.

If said yes, the name of the workspace and path to the directory is asked.

Currently, a PVC is created for the workspace, and files are copied into it.
The size of PVC is fixed in code but later can be asked as input for the user.

Approach:
- Creates a pvc with same name as a workspace (Size: 1Gi)
- Spin up an nginx pod which mount on the pvc created.
- Uses kubectl cp command (kubectl pkg) to copy the directory from the local machine
- Once copied, deletes the nginx pod.
 
Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

cc @vdemeester 

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Run the code checkers with `make check`
- [ ] Regenerate the manpages, docs and go formatting with `make generated`
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

### CODE REVIEW

**Comment by sbose78** [comment]
Could you please highlight the approach you are taking to copy the artifact over?

**Comment by sm43** [comment]
> Could you please highlight the approach you are taking to copy the artifact over?

It creates a pcc, then spins up an Nginx pod which mount on that pvc,
Once the pod is running, it uses the `kubectl cp` command to copy the directory from the local machine
Once copied, deletes the pod.

File: pkg/cmd/pipeline/start.go [hunk 1]
```diff
@@ -81,6 +82,12 @@ type startOptions struct {
 	UseParamDefaults   bool
 	TektonOptions      flags.TektonOptions
 	PodTemplate        string
+	LocalToWorkspace   []LocalToWorkspace
+}
+
+type LocalToWorkspace struct {
+	Name    string
+	DirPath string
```
Comment (vinamra28) [inline pkg/cmd/pipeline/start.go:90]: should we take storage size and access mode as well in input as AccessModes may vary depending on which platform cluster is running. Ref:- https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes

File: pkg/workspaces/local.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+func (l *LocalToWorkspace) pvcTemplate() *corev1.PersistentVolumeClaim {
+
+	return &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      l.Name,
+			Namespace: l.Namespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteMany},
```
Comment (vinamra28) [inline pkg/workspaces/local.go]: same as above. Instead of hardcoding should we not ask it from user ?

**Comment by ImJasonH** [comment]
Is there a TEP associated with this work? I think this might be something of interest to other projects (dashboard, results, chains, etc) and could benefit from having their input.

**Comment by pritidesai** [comment]
I have created an issue to reproduce this PoC in experimental repo https://github.com/tektoncd/cli/issues/1399

**Pull request closed without merging**